### PR TITLE
Publish multiple locales 

### PR DESCRIPTION
**Pull request by jhoward1994**
### What does it do?

Adds bulk locale publish modal to the CM edit view

### Why is it needed?

V5 core feature

### Related issue(s)/PR(s)

CONTENT-1903
DX-1335
DX-1364

See:
https://github.com/strapi/strapi/pull/20046 (backend)
https://github.com/strapi/strapi/pull/20069 (frontend)

### CODE REVIEW

File: packages/core/content-manager/server/src/services/document-metadata.ts [hunk 1]
```diff
@@ -79,26 +89,57 @@ export default ({ strapi }: { strapi: Core.Strapi }) => ({
...
-    return (
-      Object.values(versionsByLocale)
-        .map((localeVersions: DocumentVersion[]) => {
-          // There will not be a draft and a version counterpart if the content type does not have draft and publish
-          if (!contentTypes.hasDraftAndPublish(strapi.getModel(uid))) {
-            return pick(AVAILABLE_LOCALES_FIELDS, localeVersions[0]);
+    // There will not be a draft and a version counterpart if the content
+    // type does not have draft and publish
+    const model = strapi.getModel(uid);
+    const keysToKeep = [...AVAILABLE_LOCALES_FIELDS, ...validatableFields];
```
Comment (Marc-Roig) [inline packages/core/content-manager/server/src/services/document-metadata.ts:95]: how does the validatableFields string[] works on nested schemas? what happens if the root has a "name" field, but a component inside also has a name field?

**Comment by Marc-Roig** [comment]
As discussed, I still feel uncomfortable by coupling bulk actions frontend validation into the document metadata. And I still think we would benefit from loading the document metadata asynchronously on the CM, and not returning it on the main entry retrieval requests.

But that might be a bit of work, so let's explore first this solution and we can refactor later (famous last words)

File: packages/core/content-manager/server/src/services/document-metadata.ts (hunk 1 above)
Comment (jhoward1994) [reply packages/core/content-manager/server/src/services/document-metadata.ts:95]: I think this is covered in this test

https://github.com/strapi/strapi/blob/483791373ac11e4d33672122e16626f3104957e8/tests/api/core/content-manager/api/document-metadata.test.api.js#L266-L267

Both the product CT and features components have a 'name' string field

https://github.com/strapi/strapi/blob/483791373ac11e4d33672122e16626f3104957e8/packages/core/content-manager/server/src/services/document-metadata.ts#L230-L231
The populate here becomes 

```
	{
	  name: true,
	  category: true,
	  price: true,
	  features: {
	    populate: {
	    },
	  },
	}
```

The keys of which are used to build `keysToKeep`. So `getAvailableLocales` returns

```
	{
	  id: 2,
	  documentId: "bpd5tcr2mngn2j0e15jjakq9",
	  name: "prod-fr-draft",
	  category: "Cat-1-fr",
	  price: 1,
	  createdAt: "2024-05-10T11:23:53.441Z",
	  updatedAt: "2024-05-10T11:23:53.441Z",
	  publishedAt: null,
	  locale: "fr",
	  features: {
	    id: 2,
	    name: "Feature 1 fr",
	  },
	  status: "draft",
	}
```

Comment (Marc-Roig) [reply packages/core/content-manager/server/src/services/document-metadata.ts:95]: Probably not a big deal, but this might return more atteibutes than necessary. But I see!

**Review by Marc-Roig — APPROVED** [review]
Code looks good, didnt QA though

File: packages/core/content-manager/admin/src/pages/ListView/components/BulkActions/PublishAction.tsx [hunk 2]
```diff
@@ -324,7 +324,8 @@ const SelectedEntriesModalContent = ({
         }),
       });
 
-      setPublishedCount(res.data.count);
+      // TODO fix for V5
```
Comment (joshuaellis) [inline packages/core/content-manager/admin/src/pages/ListView/components/BulkActions/PublishAction.tsx:327]: .... this is v5?

File: packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+type Status = Modules.Documents.Params.PublicationStatus.Kind | 'modified';
+
+/* -------------------------------------------------------------------------------------------------
+ * EntryValidationText
+ * -----------------------------------------------------------------------------------------------*/
+
+const TypographyMaxWidth = styled(Typography)`
+  max-width: 300px;
+`;
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx:26]: max width can be applied to the component, you dont need this.

File: packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+  if (validationErrors) {
+    const validationErrorsMessages = Object.entries(validationErrors)
+      .map(([key, value]) => `${key}: ${formatMessage(value)}`)
+      .join(' ');
+
+    return (
+      <Flex gap={2}>
+        <CrossCircle fill="danger600" />
+        <Tooltip description={validationErrorsMessages}>
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx:44]:
```suggestion
        <Tooltip label={validationErrorsMessages}>
```

File: packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+  headers,
+  rows,
+  localesMetadata,
+  onClose,
+  validationErrors = {},
+}: BulkLocaleActionModalProps) => {
+  const { formatMessage } = useIntl();
+  const navigate = useNavigate();
+
+  const selectedRows: LocaleStatus[] = useTable<LocaleStatus[]>(
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx:128]:
```suggestion
  const selectedRows = useTable<LocaleStatus[]>(
```

File: packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+                      label={formatMessage(
+                        {
+                          id: getTranslation('Settings.list.actions.edit'),
+                          defaultMessage: 'Edit {name} locale',
+                        },
+                        {
+                          name: locale,
+                        }
+                      )}
+                      icon={<Pencil />}
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx:238]: i'd make this `children` instead. we remove the `icon` prop in the next DS beta

File: packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+                      </Status>
+                    </Box>
+                  </Table.Cell>
+                  <Table.Cell>
+                    <EntryValidationText validationErrors={error} status={status} />
+                  </Table.Cell>
+                  <Table.Cell>
+                    <IconButton
+                      onClick={() => {
+                        navigateToLocale(locale);
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx:227]: why a button not an `anchor`? if we're navigating somewhere.

File: packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx [hunk 8]
```diff
@@ -226,6 +237,299 @@ const DeleteLocaleAction: DocumentActionComponent = ({
...
+    {
+      label: formatMessage({
+        id: getTranslation('CMEditViewBulkLocale.publication-status'),
+        defaultMessage: 'Publication Status',
+      }),
+      name: 'publication-status',
+    },
+  ];
+
+  const [rows, validationErrors] = React.useMemo(() => {
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx:311]: i find this whole block really hard to follow, can we add some comments to explain why we're doing things?

File: packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx [hunk 9]
```diff
@@ -226,6 +237,299 @@ const DeleteLocaleAction: DocumentActionComponent = ({
...
+    }
+
+    const rowsFromMeta: LocaleStatus[] = documentMeta?.availableLocales.map((doc) => {
+      const { locale, status } = doc;
+
+      return { locale, status };
+    });
+    rowsFromMeta.unshift({
+      locale: document?.locale ?? '',
+      status: document?.status ?? 'draft',
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx:323]: is document undefined here? I think above you check it, so you probs dont need the optional chaining and fallbacks?

File: packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx [hunk 10]
```diff
@@ -226,6 +237,299 @@ const DeleteLocaleAction: DocumentActionComponent = ({
...
+        return document;
+      }
+
+      const validation = validate(document as Modules.Documents.AnyDocument);
+      if (validation !== null) {
+        errors[document.locale] = validation;
+      }
+
+      return document;
+    });
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx:344]:
```suggestion
    const errors = allDocuments.reduce<Record<
      Modules.Documents.Params.Locale.StringNotation,
      Record<string, MessageDescriptor>
    >>((errs, document) => {
      if (!document) {
        return errs;
      }

      const validation = validate(document as Modules.Documents.AnyDocument);
      if (validation !== null) {
        errs[document.locale] = validation;
      }

      return errs;
    }, {});
```

I think this would be a lot clearer as a reduce function:

File: packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx [hunk 11]
```diff
@@ -226,6 +237,299 @@ const DeleteLocaleAction: DocumentActionComponent = ({
...
+
+    return [rowsFromMeta, errors];
+  }, [document, documentMeta?.availableLocales, validate]);
+
+  const localesToPublish = selectedRows
+    .filter(
+      (selectedRow) =>
+        selectedRow.status !== 'published' &&
+        !Object.keys(validationErrors || {}).includes(selectedRow.locale)
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx:353]:
```suggestion
        !Object.keys(validationErrors).includes(selectedRow.locale)
```

validationErrors is always an object i think?

File: packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx [hunk 12]
```diff
@@ -226,6 +237,299 @@ const DeleteLocaleAction: DocumentActionComponent = ({
...
+        errors[document.locale] = validation;
+      }
+
+      return document;
+    });
+
+    return [rowsFromMeta, errors];
+  }, [document, documentMeta?.availableLocales, validate]);
+
+  const localesToPublish = selectedRows
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx:349]: this is optional, but instead of doing a filter, then a map, you could just do a `reduce`. I don't think it'll be an issue here, because no ones gonna have 1000s of locales....

File: packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx [hunk 13]
```diff
@@ -226,6 +237,299 @@ const DeleteLocaleAction: DocumentActionComponent = ({
...
+        },
+      });
+
+      if ('error' in bulkPublishRes) {
+        toggleNotification({
+          type: 'danger',
+          message: formatAPIError(bulkPublishRes.error),
+        });
+        return;
+      }
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx:413]: Is this not handled in the `usedocumentActions` hook?

File: packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx [hunk 14]
```diff
@@ -226,6 +237,299 @@ const DeleteLocaleAction: DocumentActionComponent = ({
...
+          setIsConfirmationOpen(false);
+        },
+        type: 'dialog',
+        title: formatMessage({
+          id: getTranslation('actions.publish.dialog.title'),
+          defaultMessage: 'Confirmation',
+        }),
+        content: (
+          <Flex direction="column" alignItems="center" gap={2}>
+            <WarningCircle width="24px" height="24px" fill="danger600" />
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx:464]:
```suggestion
            <WarningCircle width="2.4rem" height="2.4rem" fill="danger600" />
```

rem, not px 👍🏼

**Review by joshuaellis — COMMENTED** [review]
some refinement points 🧽

**Review by joshuaellis — APPROVED** [review]
